File: javascript/src/totp.js
/**
 * Shield TOTP - Time-based One-Time Passwords (RFC 6238).
 *
 * Compatible with Google Authenticator, Authy, Microsoft Authenticator, etc.
 */

const crypto = require('crypto');

/**
 * TOTP generator/verifier.
 */
class TOTP {
    /**
     * Initialize TOTP generator.
     *
     * @param {Buffer} secret - Shared secret (typically 20 bytes)
     * @param {Object} options - Optional settings
     * @param {number} options.digits - OTP length (6 or 8)
     * @param {number} options.interval - Time step in seconds
     * @param {string} options.algorithm - 'sha1' (compatible) or 'sha256'
     */
    constructor(secret, options = {}) {
        this.secret = secret;
        this.digits = options.digits || 6;
        this.interval = options.interval || 30;
        this.algorithm = options.algorithm || 'sha1';
    }

    /**
     * Generate random secret for new 2FA setup.
     *
     * @param {number} length - Secret length in bytes
     * @returns {Buffer}
     */
    static generateSecret(length = 20) {
        return crypto.randomBytes(length);
    }

    /**
     * Convert secret to base32 for QR codes.
     *
     * @param {Buffer} secret - Secret bytes
     * @returns {string} Base32 encoded string (without padding)
     */
    static secretToBase32(secret) {
        const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
        let bits = '';
        for (const byte of secret) {
            bits += byte.toString(2).padStart(8, '0');
        }
        let result = '';
        for (let i = 0; i < bits.length; i += 5) {
            const chunk = bits.substr(i, 5).padEnd(5, '0');
            result += alphabet[parseInt(chunk, 2)];
        }
        return result;
    }

    /**
     * Parse base32 secret from authenticator app.
     *
     * @param {string} b32 - Base32 encoded secret
     * @returns {Buffer} Secret bytes
     */
    static secretFromBase32(b32) {
        const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
        let bits = '';
        for (const char of b32.toUpperCase().replace(/=/g, '')) {
            const val = alphabet.indexOf(char);
            if (val >= 0) bits += val.toString(2).padStart(5, '0');
        }
        const bytes = [];
        for (let i = 0; i + 8 <= bits.length; i += 8) {
            bytes.push(parseInt(bits.substr(i, 8), 2));
        }
        return Buffer.from(bytes);
    }

    /**
     * Generate current TOTP code.
     *
     * @param {number} timestamp - Unix timestamp (default: current time)
     * @returns {string} OTP code
     */
    generate(timestamp = null) {
        if (timestamp === null) {
            timestamp = Math.floor(Date.now() / 1000);
        }
        const counter = Math.floor(timestamp / this.interval);
        return this._hotp(counter);
    }

    /**
     * Verify TOTP code with time window.
     *
     * @param {string} code - User-provided code
     * @param {number} timestamp - Time to verify against
     * @param {number} window - Number of intervals to check before/after
     * @returns {boolean}
     */
    verify(code, timestamp = null, window = 1) {
        if (timestamp === null) {
            timestamp = Math.floor(Date.now() / 1000);
        }
        const counter = Math.floor(timestamp / this.interval);

        for (let offset = -window; offset <= window; offset++) {
            const expected = this._hotp(counter + offset);
            if (crypto.timingSafeEqual(
                Buffer.from(code),
                Buffer.from(expected)
            )) {
                return true;
            }
        }
        return false;
    }

    _hotp(counter) {
        // Counter as 8-byte big-endian
        const counterBytes = Buffer.alloc(8);
        counterBytes.writeBigUInt64BE(BigInt(counter));

        // HMAC
        const h = crypto.createHmac(this.algorithm, this.secret)
            .update(counterBytes)
            .digest();

        // Dynamic truncation
        const offset = h[h.length - 1] & 0x0F;
        const code = (h.readUInt32BE(offset) & 0x7FFFFFFF) %
            Math.pow(10, this.digits);

        return code.toString().padStart(this.digits, '0');
    }

    /**
     * Generate URI for QR code.
     *
     * @param {string} account - User account identifier
     * @param {string} issuer - Service name
     * @returns {string} otpauth:// URI
     */
    provisioningUri(account, issuer = 'Shield') {
        const secret = TOTP.secretToBase32(this.secret);
        return `otpauth://totp/${issuer}:${account}` +
            `?secret=${secret}&issuer=${issuer}` +
            `&algorithm=${this.algorithm.toUpperCase()}&digits=${this.digits}`;
    }
}

/**
 * Recovery codes for 2FA backup.
 */
class RecoveryCodes {
    /**
     * Initialize with existing codes or generate new ones.
     *
     * @param {string[]} codes - List of existing codes
     */
    constructor(codes = null) {
        if (codes === null) {
            codes = RecoveryCodes.generateCodes();
        }
        this._codes = new Set(codes);
        this._used = new Set();
    }

    /**
     * Generate recovery codes.
     *
     * @param {number} count - Number of codes
     * @param {number} length - Length of each code
     * @returns {string[]}
     */
    static generateCodes(count = 10, length = 8) {
        const codes = [];
        for (let i = 0; i < count; i++) {
            const code = crypto.randomBytes(length / 2).toString('hex').toUpperCase();
            codes.push(`${code.slice(0, 4)}-${code.slice(4)}`);
        }
        return codes;
    }

    /**
     * Verify and consume a recovery code.
     *
     * @param {string} code - Recovery code to verify
     * @returns {boolean}
     */
    verify(code) {
        // Normalize format
        let normalized = code.toUpperCase().replace(/-/g, '').replace(/ /g, '');
        if (normalized.length === 8) {
            normalized = `${normalized.slice(0, 4)}-${normalized.slice(4)}`;
        }

        if (this._codes.has(normalized) && !this._used.has(normalized)) {
            this._used.add(normalized);
            return true;
        }
        return false;
    }

    /**
     * Number of unused recovery codes.
     * @returns {number}
     */
    get remaining() {
        return this._codes.size - this._used.size;
    }

    /**
     * Get all recovery codes.
     * @returns {string[]}
     */
    get codes() {
        return Array.from(this._codes).sort();
    }
}

module.exports = { TOTP, RecoveryCodes };

File: ios/Sources/Shield/TOTP.swift
import Foundation
import CommonCrypto

/// Time-based One-Time Password generator/verifier (RFC 6238).
///
/// Compatible with Google Authenticator, Authy, Microsoft Authenticator, etc.
///
/// Example:
/// ```swift
/// let secret = TOTP.generateSecret()
/// let totp = TOTP(secret: secret)
/// let code = totp.generate()
/// totp.verify(code) // true
/// ```
public final class TOTP {

    // MARK: - Types

    /// Supported HMAC algorithms.
    public enum Algorithm: String {
        case sha1 = "SHA1"
        case sha256 = "SHA256"

        var ccAlgorithm: CCHmacAlgorithm {
            switch self {
            case .sha1: return CCHmacAlgorithm(kCCHmacAlgSHA1)
            case .sha256: return CCHmacAlgorithm(kCCHmacAlgSHA256)
            }
        }

        var digestLength: Int {
            switch self {
            case .sha1: return Int(CC_SHA1_DIGEST_LENGTH)
            case .sha256: return Int(CC_SHA256_DIGEST_LENGTH)
            }
        }
    }

    // MARK: - Properties

    private let secret: [UInt8]
    private let digits: Int
    private let interval: Int
    private let algorithm: Algorithm

    // MARK: - Initialization

    /// Initialize TOTP generator.
    ///
    /// - Parameters:
    ///   - secret: Shared secret (typically 20 bytes)
    ///   - digits: OTP length (6 or 8)
    ///   - interval: Time step in seconds (default: 30)
    ///   - algorithm: `.sha1` (compatible) or `.sha256` (stronger)
    public init(
        secret: [UInt8],
        digits: Int = 6,
        interval: Int = 30,
        algorithm: Algorithm = .sha1
    ) {
        self.secret = secret
        self.digits = digits
        self.interval = interval
        self.algorithm = algorithm
    }

    // MARK: - Static Methods

    /// Generate random secret for new 2FA setup.
    ///
    /// - Parameter length: Secret length in bytes (default: 20)
    /// - Returns: Random secret bytes
    public static func generateSecret(length: Int = 20) -> [UInt8] {
        var bytes = [UInt8](repeating: 0, count: length)
        _ = SecRandomCopyBytes(kSecRandomDefault, length, &bytes)
        return bytes
    }

    /// Convert secret to base32 for QR codes.
    ///
    /// - Parameter secret: Secret bytes
    /// - Returns: Base32 encoded string (without padding)
    public static func secretToBase32(_ secret: [UInt8]) -> String {
        base32Encode(secret).trimmingCharacters(in: CharacterSet(charactersIn: "="))
    }

    /// Parse base32 secret from authenticator app.
    ///
    /// - Parameter b32: Base32 encoded secret
    /// - Returns: Secret bytes
    public static func secretFromBase32(_ b32: String) -> [UInt8] {
        // Add padding if needed
        var padded = b32.uppercased()
        let padding = (8 - (padded.count % 8)) % 8
        padded += String(repeating: "=", count: padding)
        return base32Decode(padded)
    }

    // MARK: - Public Methods

    /// Generate current TOTP code.
    ///
    /// - Parameter timestamp: Unix timestamp in seconds (default: current time)
    /// - Returns: OTP code as string (zero-padded)
    public func generate(timestamp: Int? = nil) -> String {
        let time = timestamp ?? Int(Date().timeIntervalSince1970)
        let counter = time / interval
        return hotp(counter: counter)
    }

    /// Verify TOTP code with time window.
    ///
    /// - Parameters:
    ///   - code: User-provided code
    ///   - timestamp: Time to verify against (default: now)
    ///   - window: Number of intervals to check before/after (default: 1)
    /// - Returns: True if code is valid
    public func verify(_ code: String, timestamp: Int? = nil, window: Int = 1) -> Bool {
        let time = timestamp ?? Int(Date().timeIntervalSince1970)
        let counter = time / interval

        // Check current and adjacent intervals (handles clock skew)
        for offset in -window...window {
            let expected = hotp(counter: counter + offset)
            if constantTimeEquals(code, expected) {
                return true
            }
        }
        return false
    }

    /// Generate URI for QR code (otpauth://).
    ///
    /// - Parameters:
    ///   - account: User account identifier (e.g., email)
    ///   - issuer: Service name (default: "Shield")
    /// - Returns: otpauth:// URI for QR code generation
    public func provisioningUri(account: String, issuer: String = "Shield") -> String {
        let secretB32 = Self.secretToBase32(secret)
        return "otpauth://totp/\(issuer):\(account)" +
               "?secret=\(secretB32)&issuer=\(issuer)" +
               "&algorithm=\(algorithm.rawValue)&digits=\(digits)"
    }

    // MARK: - Private Methods

    /// HOTP algorithm (RFC 4226).
    private func hotp(counter: Int) -> String {
        // Counter as 8-byte big-endian
        var counterValue = UInt64(counter).bigEndian
        let counterBytes = withUnsafeBytes(of: &counterValue) { Array($0) }

        // HMAC
        var hmacResult = [UInt8](repeating: 0, count: algorithm.digestLength)
        CCHmac(algorithm.ccAlgorithm, secret, secret.count, counterBytes, counterBytes.count, &hmacResult)

        // Dynamic truncation
        let offset = Int(hmacResult[hmacResult.count - 1] & 0x0F)
        let codeInt = (UInt32(hmacResult[offset] & 0x7F) << 24) |
                      (UInt32(hmacResult[offset + 1] & 0xFF) << 16) |
                      (UInt32(hmacResult[offset + 2] & 0xFF) << 8) |
                      UInt32(hmacResult[offset + 3] & 0xFF)

        // Modulo to get digits
        let divisor = UInt32(pow(10.0, Double(digits)))
        let code = codeInt % divisor

        return String(format: "%0\(digits)d", code)
    }

    private func constantTimeEquals(_ a: String, _ b: String) -> Bool {
        guard a.count == b.count else { return false }
        var result: UInt8 = 0
        for (charA, charB) in zip(a.utf8, b.utf8) {
            result |= charA ^ charB
        }
        return result == 0
    }

    // MARK: - Base32 Encoding

    private static let base32Alphabet = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZ234567")

    private static func base32Encode(_ data: [UInt8]) -> String {
        guard !data.isEmpty else { return "" }
        var result = ""
        var buffer = 0
        var bitsLeft = 0

        for byte in data {
            buffer = (buffer << 8) | Int(byte)
            bitsLeft += 8
            while bitsLeft >= 5 {
                let index = (buffer >> (bitsLeft - 5)) & 0x1F
                result.append(base32Alphabet[index])
                bitsLeft -= 5
            }
        }

        if bitsLeft > 0 {
            let index = (buffer << (5 - bitsLeft)) & 0x1F
            result.append(base32Alphabet[index])
        }

        // Add padding
        while result.count % 8 != 0 {
            result.append("=")
        }

        return result
    }

    private static func base32Decode(_ data: String) -> [UInt8] {
        let cleanData = data.uppercased().trimmingCharacters(in: CharacterSet(charactersIn: "="))
        guard !cleanData.isEmpty else { return [] }

        var result = [UInt8]()
        var buffer = 0
        var bitsLeft = 0

        for char in cleanData {
            guard let index = base32Alphabet.firstIndex(of: char) else { continue }
            buffer = (buffer << 5) | index
            bitsLeft += 5
            if bitsLeft >= 8 {
                result.append(UInt8(buffer >> (bitsLeft - 8)))
                bitsLeft -= 8
            }
        }

        return result
    }
}

// MARK: - RecoveryCodes

/// Recovery codes for 2FA backup.
///
/// Use when user loses access to their authenticator app.
/// Each code can only be used once.
///
/// Example:
/// ```swift
/// let recovery = RecoveryCodes()
/// print(recovery.codes) // Show codes to user
/// recovery.verify("ABCD-1234") // Consumes the code
/// ```
public final class RecoveryCodes {

    private var _codes: Set<String>
    private var _used: Set<String> = []

    /// Number of unused recovery codes.
    public var remaining: Int {
        _codes.count - _used.count
    }

    /// Get all recovery codes (for display to user).
    public var codes: [String] {
        _codes.sorted()
    }

    /// Initialize with existing codes or generate new ones.
    ///
    /// - Parameter codes: List of existing codes, or nil to generate new
    public init(codes: [String]? = nil) {
        _codes = Set(codes ?? Self.generateCodes())
    }

    /// Generate recovery codes.
    ///
    /// - Parameters:
    ///   - count: Number of codes to generate (default: 10)
    ///   - length: Length of each code in hex chars (default: 8)
    /// - Returns: List of recovery codes
    public static func generateCodes(count: Int = 10, length: Int = 8) -> [String] {
        (0..<count).map { _ in
            var bytes = [UInt8](repeating: 0, count: length / 2)
            _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
            let hex = bytes.map { String(format: "%02X", $0) }.joined()
            return "\(hex.prefix(4))-\(hex.suffix(4))"
        }
    }

    /// Verify and consume a recovery code.
    ///
    /// - Parameter code: Recovery code to verify
    /// - Returns: True if valid (code is now consumed)
    public func verify(_ code: String) -> Bool {
        // Normalize format
        let normalized = code.uppercased()
            .replacingOccurrences(of: "-", with: "")
            .replacingOccurrences(of: " ", with: "")

        let formatted: String
        if normalized.count == 8 {
            formatted = "\(normalized.prefix(4))-\(normalized.suffix(4))"
        } else {
            formatted = code.uppercased()
        }

        if _codes.contains(formatted) && !_used.contains(formatted) {
            _used.insert(formatted)
            return true
        }
        return false
    }
}
